zebra-chain/src/primitives/proofs/halo2.rs

```diff
@@ -0,0 +1,77 @@
+use serde::{Deserialize, Serialize};
+use std::{fmt, io};
+
+use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
+
+/// An encoding of a Halo2 proof, as used in [Zcash][halo2].
+///
+/// Halo2 proofs in Zcash Orchard do not have a fixed size, hence the newtype
+/// around a vector of bytes.
+///
+/// [halo2]: https://zips.z.cash/protocol/nu5.pdf#halo2
+#[derive(Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Halo2Proof(pub Vec<u8>);
+
+impl fmt::Debug for Halo2Proof {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("Halo2Proof")
+            .field(&hex::encode(&self.0[..]))
+            .finish()
+    }
+}
+
+// These impls all only exist because of array length restrictions.
```

Comment:
```suggestion
// These impls all only exist because of array length restrictions.
// TODO: use const generics https://github.com/ZcashFoundation/zebra/issues/2042
```